server/schedule/region_scatterer.go

```diff
@@ -346,74 +286,60 @@ func (r *RegionScatterer) scatterRegion(region *core.RegionInfo, group string) *
 	op.SetPriorityLevel(core.HighPriority)
 	return op
 }
 
-func (r *RegionScatterer) selectPeerToReplace(group string, stores map[uint64]*core.StoreInfo, region *core.RegionInfo, oldPeer *metapb.Peer, context engineContext) *metapb.Peer {
-	// scoreGuard guarantees that the distinct score will not decrease.
-	storeID := oldPeer.GetStoreId()
-	sourceStore := r.cluster.GetStore(storeID)
+func (r *RegionScatterer) selectCandidates(region *core.RegionInfo, sourceStoreID uint64, selectedStores map[uint64]struct{}, context engineContext) []uint64 {
+	sourceStore := r.cluster.GetStore(sourceStoreID)
 	if sourceStore == nil {
-		log.Error("failed to get the store", zap.Uint64("store-id", storeID), errs.ZapError(errs.ErrGetSourceStore))
+		log.Error("failed to get the store", zap.Uint64("store-id", sourceStoreID), errs.ZapError(errs.ErrGetSourceStore))
 		return nil
 	}
-	scoreGuard := filter.NewPlacementSafeguard(r.name, r.cluster, region, sourceStore)
-
-	candidates := make([]*core.StoreInfo, 0, len(stores))
-	for _, store := range stores {
-		if !scoreGuard.Target(r.cluster.GetOpts(), store) {
-			continue
-		}
-		candidates = append(candidates, store)
-	}
-
-	if len(candidates) == 0 {
-		return nil
-	}
-
-	minPeer := uint64(math.MaxUint64)
-	var selectedCandidateID uint64
-	for _, candidate := range candidates {
-		count := context.selectedPeer.get(candidate.GetID(), group)
-		if count < minPeer {
-			minPeer = count
-			selectedCandidateID = candidate.GetID()
-		}
-	}
-	if selectedCandidateID < 1 {
-		target := candidates[rand.Intn(len(candidates))]
-		return &metapb.Peer{
-			StoreId: target.GetID(),
-			Role:    oldPeer.GetRole(),
-		}
-	}
-
-	return &metapb.Peer{
-		StoreId: selectedCandidateID,
-		Role:    oldPeer.GetRole(),
-	}
-}
-
-func (r *RegionScatterer) collectAvailableStores(group string, region *core.RegionInfo, context engineContext) map[uint64]*core.StoreInfo {
 	filters := []filter.Filter{
-		filter.NewExcludedFilter(r.name, nil, region.GetStoreIds()),
-		&filter.StoreStateFilter{ActionScope: r.name, MoveRegion: true},
+		filter.NewExcludedFilter("scatter-region", nil, selectedStores),
 	}
+	scoreGuard := filter.NewPlacementSafeguard(r.name, r.cluster, region, sourceStore)
 	filters = append(filters, context.filters...)
-	filters = append(filters, context.selectedPeer.newFilters(r.name, group)...)
-
+	filters = append(filters, scoreGuard)
 	stores := r.cluster.GetStores()
-	targets := make(map[uint64]*core.StoreInfo, len(stores))
+	candidates := make([]uint64, 0)
 	for _, store := range stores {
 		if filter.Target(r.cluster.GetOpts(), store, filters) && !store.IsBusy() {
-			targets[store.GetID()] = store
+			candidates = append(candidates, store.GetID())
+		}
+	}
+	return candidates
+}
+
+func (r *RegionScatterer) selectStore(group string, peer *metapb.Peer, candidates []uint64, context engineContext) *metapb.Peer {
+	if len(candidates) < 1 {
+		return peer
+	}
+	var newPeer *metapb.Peer
+	minCount := uint64(math.MaxUint64)
```

Comment:
should we need to compare with the count of source peer's store? and add a test case `TestFourStores ` with 1000 regions to check it.

Reply:
updated.